tensorflow/core/kernels/fft_ops.cc

```diff
@@ -97,9 +84,141 @@ class FFTGPUBase : public OpKernel {
+        // Convert output to tensor and get tensor size for slicing.
+        auto output = out->flat_inner_dims<complex64, FFTRank + 1>();
+        auto sizes = output.dimensions();
+        // Slice the full FFT to get the non-negative frequency components only.
+        output.slice(startIndices, sizes) =
+          full_fft.slice(startIndices, sizes);
+      }
+      else {
+        auto input = ((Tensor) in).flat_inner_dims<complex64, FFTRank + 1>();
+        // The first dimension contains the zero-frequency component which we
+        // do not want to duplicate. So we reconstruct the complex signal by
+        // (1) slicing from the second element, (2) reversing the order,
+        // (3) taking the complex conjugate, (4) concatenating with the original
+        // input. Note that for an even input length, the last element is the
+        // Nyquist frequency which we also do not want to duplicate.
+        Eigen::DSizes<Eigen::DenseIndex, FFTRank + 1> startIndices;
+        startIndices[FFTRank] = 1;
+        auto sizes = input.dimensions();
+        if (sizes[FFTRank] % 2 == 0) {
+          sizes[FFTRank] -= 1;
+        }
+        auto cc = input.slice(startIndices, sizes).conjugate()
+          .reverse(FFTRank);
+        auto full_fft = input.concatenate(cc, FFTRank);
+
+        // Evaluate the IFFT
+        auto output = out->flat_inner_dims<float, FFTRank + 1>();
+        FFTFunctor<CPUDevice, complex64, float, Eigen::RealPart,
+                  Eigen::FFT_REVERSE, FFTRank> functor;
+        functor(ctx->eigen_device<CPUDevice>(), output, full_fft);
+      }
+    }
+  }
+};
+
+REGISTER_KERNEL_BUILDER(Name("FFT").Device(DEVICE_CPU), FFTCPU<true, false, 1>);
+REGISTER_KERNEL_BUILDER(Name("IFFT").Device(DEVICE_CPU),
+                        FFTCPU<false, false, 1>);
+REGISTER_KERNEL_BUILDER(Name("FFT2D").Device(DEVICE_CPU),
+                        FFTCPU<true, false, 2>);
+REGISTER_KERNEL_BUILDER(Name("IFFT2D").Device(DEVICE_CPU),
+                        FFTCPU<false, false, 2>);
+REGISTER_KERNEL_BUILDER(Name("FFT3D").Device(DEVICE_CPU),
+                        FFTCPU<true, false, 3>);
+REGISTER_KERNEL_BUILDER(Name("IFFT3D").Device(DEVICE_CPU),
+                        FFTCPU<false, false, 3>);
+
+REGISTER_KERNEL_BUILDER(Name("RFFT").Device(DEVICE_CPU), FFTCPU<true, true, 1>);
+REGISTER_KERNEL_BUILDER(Name("IRFFT").Device(DEVICE_CPU),
+                        FFTCPU<false, false, 1>);
+REGISTER_KERNEL_BUILDER(Name("RFFT2D").Device(DEVICE_CPU),
+                        FFTCPU<true, true, 2>);
+REGISTER_KERNEL_BUILDER(Name("IRFFT2D").Device(DEVICE_CPU),
+                        FFTCPU<false, false, 2>);
+REGISTER_KERNEL_BUILDER(Name("RFFT3D").Device(DEVICE_CPU),
+                        FFTCPU<true, true, 3>);
+REGISTER_KERNEL_BUILDER(Name("IRFFT3D").Device(DEVICE_CPU),
+                        FFTCPU<false, false, 3>);
+
+#if GOOGLE_CUDA
+#include "tensorflow/core/platform/stream_executor.h"
```

Comment:
Shouldn't we move this at the top of the file to avoid including stream_executor.h in the tensorflow namespace ?